File: python/packages/core/agent_framework/_serialization.py
# Copyright (c) Microsoft. All rights reserved.

import json
import re
from collections.abc import Mapping, MutableMapping
from typing import Any, ClassVar, Protocol, TypeVar, runtime_checkable

from ._logging import get_logger

logger = get_logger()

TClass = TypeVar("TClass", bound="SerializationMixin")
TProtocol = TypeVar("TProtocol", bound="SerializationProtocol")

# Regex pattern for converting CamelCase to snake_case
_CAMEL_TO_SNAKE_PATTERN = re.compile(r"(?<!^)(?=[A-Z])")


@runtime_checkable
class SerializationProtocol(Protocol):
    """Protocol for objects that support serialization and deserialization.

    This protocol defines the interface for objects that can be converted to and from dictionaries.

    Examples:
        .. code-block:: python

            from agent_framework import SerializationProtocol


            class MySerializable:
                def __init__(self, value: str):
                    self.value = value

                def to_dict(self, **kwargs):
                    return {"value": self.value}

                @classmethod
                def from_dict(cls, value, **kwargs):
                    return cls(value["value"])


            # Verify it implements the protocol
            assert isinstance(MySerializable("test"), SerializationProtocol)
    """

    def to_dict(self, **kwargs: Any) -> dict[str, Any]:
        """Convert the instance to a dictionary.

        Keyword Args:
            kwargs: Additional keyword arguments for serialization.

        Returns:
            Dictionary representation of the instance.
        """
        ...

    @classmethod
    def from_dict(cls: type[TProtocol], value: MutableMapping[str, Any], /, **kwargs: Any) -> TProtocol:
        """Create an instance from a dictionary.

        Args:
            value: Dictionary containing the instance data (positional-only).

        Keyword Args:
            kwargs: Additional keyword arguments for deserialization.

        Returns:
            New instance of the class.
        """
        ...


def is_serializable(value: Any) -> bool:
    """Check if a value is JSON serializable.

    Args:
        value: The value to check.

    Returns:
        True if the value is JSON serializable, False otherwise.
    """
    return isinstance(value, (str, int, float, bool, type(None), list, dict))


class SerializationMixin:
    """Mixin class providing serialization and deserialization capabilities.

    Classes using this mixin should handle MutableMapping inputs in their __init__ method
    for any parameters that expect SerializationMixin/SerializationProtocol instances.
    The __init__ should check if the value is a MutableMapping and call from_dict() to convert it.

    So take the two classes below as an example. The first purely uses base types, strings in this case.
    The second has a param that is of the type of the first class.
    Because we setup the __init__ method to handle MutableMapping,
    we can pass in a dict to the second class and it will convert it to an instance of the first class.

    Examples:
        .. code-block:: python

            class SerializableMixinType(SerializationMixin):
                def __init__(self, param1: str, param2: int) -> None:
                    self.param1 = param1
                    self.param2 = param2


            class MyClass(SerializationMixin):
                def __init__(
                    self,
                    regular_param: str,
                    param: SerializableMixinType | MutableMapping[str, Any] | None = None,
                ) -> None:
                    if isinstance(param, MutableMapping):
                        self.param = self.from_dict(param)
                    else:
                        self.param = param
                    self.regular_param = regular_param


            instance = MyClass.from_dict({"regular_param": "value", "param": {"param1": "value1", "param2": 42}})

    A more complex use case involves an injectable dependency that is not serialized.
    In this case, the dependency is passed in via the dependencies parameter to from_dict/from_json.

    Examples:
        .. code-block:: python

            from library import Client


            class MyClass(SerializationMixin):
                INJECTABLE = {"client"}

    During serialization, the field listed as INJECTABLE (and also DEFAULT_EXCLUDE) will be excluded from the output.
    Then in deserialization, the dependencies dict uses a nested structure:
    ``{"<type>": {"<parameter>": value}}`` where ``<type>`` is the type identifier for the class
    (either the value of the 'type' class variable or the snake_cased class name if 'type' is not present),
    and ``<parameter>`` is the name of the injectable parameter in the __init__ method.
    For dict parameters, the structure is ``{"<type>": {"<dict-parameter>": {"<key>": value}}}``.
    For instance-specific dependencies, use ``{"<type>": {"<field>:<name>": {"<parameter>": value}}}``.
    """

    DEFAULT_EXCLUDE: ClassVar[set[str]] = set()
    INJECTABLE: ClassVar[set[str]] = set()

    def to_dict(self, *, exclude: set[str] | None = None, exclude_none: bool = True) -> dict[str, Any]:
        """Convert the instance and any nested objects to a dictionary.

        Keyword Args:
            exclude: The set of field names to exclude from serialization.
            exclude_none: Whether to exclude None values from the output. Defaults to True.

        Returns:
            Dictionary representation of the instance.
        """
        # Combine exclude sets
        combined_exclude = set(self.DEFAULT_EXCLUDE)
        if exclude:
            combined_exclude.update(exclude)
        combined_exclude.update(self.INJECTABLE)

        # Get all instance attributes
        result: dict[str, Any] = {} if "type" in combined_exclude else {"type": self._get_type_identifier()}
        for key, value in self.__dict__.items():
            if key not in combined_exclude and not key.startswith("_"):
                if exclude_none and value is None:
                    continue
                # Recursively serialize SerializationProtocol objects
                if isinstance(value, SerializationProtocol):
                    result[key] = value.to_dict(exclude=exclude, exclude_none=exclude_none)
                    continue
                # Handle lists containing SerializationProtocol objects
                if isinstance(value, list):
                    value_as_list: list[Any] = []
                    for item in value:
                        if isinstance(item, SerializationProtocol):
                            value_as_list.append(item.to_dict(exclude=exclude, exclude_none=exclude_none))
                            continue
                        if is_serializable(item):
                            value_as_list.append(item)
                            continue
                        logger.debug(
                            f"Skipping non-serializable item in list attribute '{key}' of type {type(item).__name__}"
                        )
                    result[key] = value_as_list
                    continue
                # Handle dicts containing SerializationProtocol values
                if isinstance(value, dict):
                    serialized_dict: dict[str, Any] = {}
                    for k, v in value.items():
                        if isinstance(v, SerializationProtocol):
                            serialized_dict[k] = v.to_dict(exclude=exclude, exclude_none=exclude_none)
                            continue
                        # Check if the value is JSON serializable
                        if is_serializable(v):
                            serialized_dict[k] = v
                            continue
                        logger.debug(
                            f"Skipping non-serializable value for key '{k}' in dict attribute '{key}' "
                            f"of type {type(v).__name__}"
                        )
                    result[key] = serialized_dict
                    continue
                # Directly include JSON serializable values
                if is_serializable(value):
                    result[key] = value
                    continue
                logger.debug(f"Skipping non-serializable attribute '{key}' of type {type(value).__name__}")

        return result

    def to_json(self, *, exclude: set[str] | None = None, exclude_none: bool = True, **kwargs: Any) -> str:
        """Convert the instance to a JSON string.

        Keyword Args:
            exclude: The set of field names to exclude from serialization.
            exclude_none: Whether to exclude None values from the output. Defaults to True.
            **kwargs: passed through to the json.dumps method.

        Returns:
            JSON string representation of the instance.
        """
        return json.dumps(self.to_dict(exclude=exclude, exclude_none=exclude_none), **kwargs)

    @classmethod
    def from_dict(
        cls: type[TClass], value: MutableMapping[str, Any], /, *, dependencies: MutableMapping[str, Any] | None = None
    ) -> TClass:
        """Create an instance from a dictionary.

        Args:
            value: The dictionary containing the instance data (positional-only).

        Keyword Args:
            dependencies: A nested dictionary mapping type identifiers to their injectable dependencies.
                Structure: ``{"<type>": {"<parameter>": value, ...}}`` for simple parameters,
                ``{"<type>": {"<dict-parameter>": {"<key>": value}}}`` for dict parameters,
                or ``{"<type>": {"<instance-key>": {"<parameter>": value}}}`` for instance-specific dependencies.

        Returns:
            New instance of the class.

        Examples:
            Simple parameter injection:

            .. code-block:: python

                class MyClass(SerializationMixin):
                    INJECTABLE = {"client"}

                    def __init__(self, value: str, client=None):
                        self.value = value
                        self.client = client


                data = {"type": "my_class", "value": "test"}
                dependencies = {"my_class": {"client": client_instance}}
                obj = MyClass.from_dict(data, dependencies=dependencies)

            Dict parameter injection:

            .. code-block:: python

                class ConfigClass(SerializationMixin):
                    INJECTABLE = {"config"}

                    def __init__(self, name: str, config: dict | None = None):
                        self.name = name
                        self.config = config or {}


                data = {"type": "config_class", "name": "app"}
                dependencies = {"config_class": {"config": {"api_key": "secret", "timeout": 30}}}
                obj = ConfigClass.from_dict(data, dependencies=dependencies)

            Instance-specific injection (when multiple instances of same type need different dependencies):

            .. code-block:: python

                class Worker(SerializationMixin):
                    INJECTABLE = {"func"}

                    def __init__(self, name: str, func=None):
                        self.name = name
                        self.func = func


                data = {"type": "worker", "name": "worker1"}
                dependencies = {"worker": {"name:worker1": {"func": my_function}}}
                obj = Worker.from_dict(data, dependencies=dependencies)
        """
        if dependencies is None:
            dependencies = {}

        # Get the type identifier
        type_id = cls._get_type_identifier(value)

        if (supplied_type := value.get("type")) and supplied_type != type_id:
            raise ValueError(f"Type mismatch: expected '{type_id}', got '{supplied_type}'")

        # Create a copy of the value dict to work with, filtering out the 'type' key
        kwargs = {k: v for k, v in value.items() if k != "type"}

        # Process dependencies using dict-based structure
        type_deps = dependencies.get(type_id, {})
        for dep_key, dep_value in type_deps.items():
            # Check if this is an instance-specific dependency (field:name format)
            if ":" in dep_key:
                field, name = dep_key.split(":", 1)
                # Only apply if the instance matches
                if kwargs.get(field) == name and isinstance(dep_value, dict):
                    # Apply instance-specific dependencies
                    for param_name, param_value in dep_value.items():
                        if param_name not in cls.INJECTABLE:
                            logger.debug(
                                f"Dependency '{param_name}' for type '{type_id}' is not in INJECTABLE set. "
                                f"Available injectable parameters: {cls.INJECTABLE}"
                            )
                        # Handle nested dict parameters
                        if (
                            isinstance(param_value, dict)
                            and param_name in kwargs
                            and isinstance(kwargs[param_name], dict)
                        ):
                            kwargs[param_name].update(param_value)
                        else:
                            kwargs[param_name] = param_value
            else:
                # Regular parameter dependency
                if dep_key not in cls.INJECTABLE:
                    logger.debug(
                        f"Dependency '{dep_key}' for type '{type_id}' is not in INJECTABLE set. "
                        f"Available injectable parameters: {cls.INJECTABLE}"
                    )
                # Handle dict parameters - merge if both are dicts
                if isinstance(dep_value, dict) and dep_key in kwargs and isinstance(kwargs[dep_key], dict):
                    kwargs[dep_key].update(dep_value)
                else:
                    kwargs[dep_key] = dep_value

        return cls(**kwargs)

    @classmethod
    def from_json(cls: type[TClass], value: str, /, *, dependencies: MutableMapping[str, Any] | None = None) -> TClass:
        """Create an instance from a JSON string.

        Args:
            value: The JSON string containing the instance data (positional-only).

        Keyword Args:
            dependencies: A nested dictionary mapping type identifiers to their injectable dependencies.
                See :meth:`from_dict` for structure and examples.

        Returns:
            New instance of the class.
        """
        data = json.loads(value)
        return cls.from_dict(data, dependencies=dependencies)

    @classmethod
    def _get_type_identifier(cls, value: Mapping[str, Any] | None = None) -> str:
        """Get the type identifier for this class.

        Returns the value of the ``type`` class variable if present,
        otherwise returns a snake_cased version of the class name.

        Returns:
            Type identifier string.
        """
        # for from_dict
        if value and (type_ := value.get("type")) and isinstance(type_, str):
            return type_  # type:ignore[no-any-return]
        # for todict when defined per instance
        if (type_ := getattr(cls, "type", None)) and isinstance(type_, str):
            return type_  # type:ignore[no-any-return]
        # for both when defined on class.
        if (type_ := getattr(cls, "TYPE", None)) and isinstance(type_, str):
            return type_  # type:ignore[no-any-return]
        # fallback and default
        # Convert class name to snake_case
        return _CAMEL_TO_SNAKE_PATTERN.sub("_", cls.__name__).lower()

File: typescript/src/core/serialization.ts
/**
 * Serialization utilities for converting objects to/from JSON with dependency injection support.
 *
 * This module provides a generic serialization framework that supports:
 * - Automatic field inclusion/exclusion based on metadata
 * - Recursive serialization of nested objects
 * - Dependency injection for non-serializable components
 * - Type identification for deserialization
 *
 * @module serialization
 */

import { getLogger } from './logging/index.js';

const logger = getLogger();

/**
 * Regular expression pattern for converting CamelCase to snake_case.
 */
const CAMEL_TO_SNAKE_PATTERN = /(?<!^)(?=[A-Z])/;

/**
 * Protocol for objects that support serialization and deserialization.
 *
 * Classes implementing this protocol can be converted to/from plain JavaScript objects
 * for JSON serialization and persistence.
 *
 * @example
 * ```typescript
 * class MyClass implements SerializationProtocol {
 *   constructor(public value: string) {}
 *
 *   toDict(): Record<string, unknown> {
 *     return { type: 'my_class', value: this.value };
 *   }
 *
 *   static fromDict(data: Record<string, unknown>): MyClass {
 *     return new MyClass(data.value as string);
 *   }
 * }
 * ```
 */
export interface SerializationProtocol {
  /**
   * Convert the instance to a dictionary representation.
   *
   * @param options - Serialization options
   * @returns Dictionary representation of the instance
   */
  toDict(options?: SerializationOptions): Record<string, unknown>;
}

/**
 * Options for controlling serialization behavior.
 */
export interface SerializationOptions {
  /**
   * Fields to exclude from serialization.
   */
  exclude?: Set<string>;

  /**
   * Whether to exclude null and undefined values from the output.
   * @default true
   */
  excludeNone?: boolean;
}

/**
 * Options for controlling deserialization behavior.
 */
export interface DeserializationOptions {
  /**
   * Dependencies to inject during deserialization.
   *
   * Keys should be in format:
   * - `"<type>.<parameter>"` for simple dependencies
   * - `"<type>.<dict-parameter>.<key>"` for nested dict dependencies
   *
   * @example
   * ```typescript
   * {
   *   'chat_agent.chatClient': clientInstance,
   *   'chat_agent.options.timeout': 5000
   * }
   * ```
   */
  dependencies?: Record<string, unknown>;
}

/**
 * Check if a value is JSON serializable.
 *
 * Returns true for primitive types, null, arrays, and plain objects.
 * Returns false for functions, class instances, and other non-JSON types.
 *
 * @param value - The value to check
 * @returns True if the value is JSON serializable
 *
 * @example
 * ```typescript
 * isSerializable('hello') // true
 * isSerializable(42) // true
 * isSerializable([1, 2, 3]) // true
 * isSerializable({ key: 'value' }) // true
 * isSerializable(new Date()) // false
 * isSerializable(() => {}) // false
 * ```
 */
export function isSerializable(value: unknown): boolean {
  if (value === null) return true;
  if (value === undefined) return false;

  const type = typeof value;
  if (type === 'string' || type === 'number' || type === 'boolean') {
    return true;
  }

  if (Array.isArray(value)) {
    return true;
  }

  // Check for plain objects (not class instances)
  if (type === 'object' && value.constructor === Object) {
    return true;
  }

  return false;
}

/**
 * Type guard to check if a value implements SerializationProtocol.
 *
 * @param value - The value to check
 * @returns True if the value has a toDict method
 *
 * @example
 * ```typescript
 * if (isSerializationProtocol(obj)) {
 *   const dict = obj.toDict();
 * }
 * ```
 */
export function isSerializationProtocol(value: unknown): value is SerializationProtocol {
  return value !== null && typeof value === 'object' && typeof (value as SerializationProtocol).toDict === 'function';
}

/**
 * Mixin class providing default serialization and deserialization behavior.
 *
 * Classes extending this mixin automatically gain serialization capabilities with:
 * - Automatic field discovery and serialization
 * - Configurable field exclusion via DEFAULT_EXCLUDE
 * - Dependency injection support via INJECTABLE
 * - Recursive serialization of nested objects
 * - Type identification for deserialization
 *
 * @example
 * ```typescript
 * class ChatAgent extends SerializationMixin {
 *   static readonly DEFAULT_EXCLUDE = new Set(['additionalProperties']);
 *   static readonly INJECTABLE = new Set(['chatClient']);
 *
 *   constructor(
 *     public chatClient: ChatClientProtocol,
 *     public name: string,
 *     public additionalProperties: Record<string, unknown> = {}
 *   ) {
 *     super();
 *   }
 * }
 *
 * const agent = new ChatAgent(client, 'my-agent');
 * const serialized = agent.toDict(); // excludes chatClient, additionalProperties
 * const json = agent.toJson();
 *
 * const restored = ChatAgent.fromDict(serialized, {
 *   dependencies: { 'chat_agent.chatClient': client }
 * });
 * ```
 */
export class SerializationMixin implements SerializationProtocol {
  /**
   * Fields to always exclude from serialization.
   *
   * Override in subclasses to specify fields that should never be serialized.
   */
  static readonly DEFAULT_EXCLUDE: Set<string> = new Set();

  /**
   * Fields that are injectable dependencies.
   *
   * These fields are:
   * - Excluded from serialization
   * - Can be provided via the dependencies parameter in fromDict/fromJson
   *
   * Override in subclasses to specify which dependencies can be injected.
   */
  static readonly INJECTABLE: Set<string> = new Set();

  /**
   * Custom type identifier for this class.
   *
   * If not specified, the class name will be converted to snake_case.
   *
   * @example
   * ```typescript
   * class MyAgent extends SerializationMixin {
   *   static readonly type = 'custom_agent_type';
   * }
   * ```
   */
  static readonly type?: string;

  /**
   * Convert the instance to a dictionary representation.
   *
   * Automatically serializes all public instance properties, excluding:
   * - Fields in DEFAULT_EXCLUDE
   * - Fields in INJECTABLE
   * - Fields in the runtime exclude parameter
   * - Private fields (starting with underscore)
   * - Null/undefined values (when excludeNone is true)
   *
   * Recursively serializes nested SerializationProtocol objects, arrays, and plain objects.
   *
   * @param options - Serialization options
   * @returns Dictionary representation of the instance
   *
   * @example
   * ```typescript
   * const dict = agent.toDict();
   * const dictWithExtras = agent.toDict({
   *   exclude: new Set(['temporaryField']),
   *   excludeNone: false
   * });
   * ```
   */
  toDict(options: SerializationOptions = {}): Record<string, unknown> {
    const { exclude = new Set(), excludeNone = true } = options;
    const constructor = this.constructor as typeof SerializationMixin;

    // Combine exclude sets
    const combined = new Set([...constructor.DEFAULT_EXCLUDE, ...constructor.INJECTABLE, ...exclude]);

    // Start with type field (unless excluded)
    const result: Record<string, unknown> = combined.has('type') ? {} : { type: this.getTypeIdentifier() };

    // Serialize all public instance properties
    for (const [key, value] of Object.entries(this)) {
      // Skip if excluded or private
      if (combined.has(key) || key.startsWith('_')) {
        continue;
      }

      // Skip null/undefined if excludeNone
      if (excludeNone && (value === null || value === undefined)) {
        continue;
      }

      // Recursively serialize SerializationProtocol objects
      if (isSerializationProtocol(value)) {
        result[key] = value.toDict(options);
        continue;
      }

      // Handle arrays
      if (Array.isArray(value)) {
        const valueAsList: unknown[] = [];
        for (const item of value) {
          if (isSerializationProtocol(item)) {
            valueAsList.push(item.toDict(options));
            continue;
          }
          if (isSerializable(item)) {
            valueAsList.push(item);
            continue;
          }
          logger.debug(`Skipping non-serializable item in list attribute '${key}' of type ${typeof item}`);
        }
        result[key] = valueAsList;
        continue;
      }

      // Handle plain objects/maps
      if (typeof value === 'object' && value !== null && value.constructor === Object) {
        const serializedObj: Record<string, unknown> = {};
        for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
          if (isSerializationProtocol(v)) {
            serializedObj[k] = v.toDict(options);
            continue;
          }
          if (isSerializable(v)) {
            serializedObj[k] = v;
            continue;
          }
          logger.debug(
            `Skipping non-serializable value for key '${k}' in dict attribute '${key}' of type ${typeof v}`
          );
        }
        result[key] = serializedObj;
        continue;
      }

      // Include if JSON serializable
      if (isSerializable(value)) {
        result[key] = value;
        continue;
      }

      logger.debug(`Skipping non-serializable attribute '${key}' of type ${typeof value}`);
    }

    return result;
  }

  /**
   * Convert the instance to a JSON string.
   *
   * @param options - Serialization options plus JSON.stringify options (indent, replacer)
   * @returns JSON string representation of the instance
   *
   * @example
   * ```typescript
   * const json = agent.toJson();
   * const prettyJson = agent.toJson({ indent: 2 });
   * ```
   */
  toJson(
    options: SerializationOptions & { indent?: number; replacer?: (key: string, value: unknown) => unknown } = {}
  ): string {
    const { indent, replacer, ...serializationOptions } = options;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return JSON.stringify(this.toDict(serializationOptions), replacer as any, indent);
  }

  /**
   * Create an instance from a dictionary.
   *
   * Injects dependencies specified in the dependencies parameter and passes
   * remaining data to the constructor.
   *
   * @param data - Dictionary containing the instance data
   * @param options - Deserialization options including dependencies
   * @returns New instance of the class
   *
   * @example
   * ```typescript
   * const data = { name: 'my-agent', model: 'gpt-4' };
   * const agent = ChatAgent.fromDict(data, {
   *   dependencies: {
   *     'chat_agent.chatClient': clientInstance
   *   }
   * });
   * ```
   */
  static fromDict<T extends SerializationMixin>(
    this: (new (kwargs: Record<string, unknown>) => T) & typeof SerializationMixin,
    data: Record<string, unknown>,
    options: DeserializationOptions = {}
  ): T {
    const { dependencies = {} } = options;
    const typeId = this.getTypeIdentifier();

    // Filter out 'type' field
    const kwargs: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      if (key !== 'type') {
        kwargs[key] = value;
      }
    }

    // Process dependencies
    for (const [depKey, depValue] of Object.entries(dependencies)) {
      const parts = depKey.split('.');
      if (parts.length < 2) {
        continue;
      }

      const [depType, param, ...rest] = parts;

      // Skip if not for this type
      if (depType !== typeId) {
        continue;
      }

      // Log debug message if dependency is not in INJECTABLE
      if (!this.INJECTABLE.has(param)) {
        logger.debug(
          `Dependency '${param}' for type '${typeId}' is not in INJECTABLE set. ` +
            `Available injectable parameters: ${Array.from(this.INJECTABLE).join(', ')}`
        );
      }

      if (rest.length === 0) {
        // Simple dependency: <type>.<parameter>
        kwargs[param] = depValue;
      } else if (rest.length === 1) {
        // Dict dependency: <type>.<dict-parameter>.<key>
        const key = rest[0];
        if (!(param in kwargs)) {
          kwargs[param] = {};
        }
        (kwargs[param] as Record<string, unknown>)[key] = depValue;
      }
    }

    return new this(kwargs);
  }

  /**
   * Create an instance from a JSON string.
   *
   * @param json - JSON string containing the instance data
   * @param options - Deserialization options including dependencies
   * @returns New instance of the class
   *
   * @example
   * ```typescript
   * const json = '{"name":"my-agent","model":"gpt-4"}';
   * const agent = ChatAgent.fromJson(json, {
   *   dependencies: {
   *     'chat_agent.chatClient': clientInstance
   *   }
   * });
   * ```
   */
  static fromJson<T extends SerializationMixin>(
    this: (new (kwargs: Record<string, unknown>) => T) & typeof SerializationMixin,
    json: string,
    options: DeserializationOptions = {}
  ): T {
    const data = JSON.parse(json) as Record<string, unknown>;
    return this.fromDict(data, options) as T;
  }

  /**
   * Get the type identifier for this instance.
   *
   * Uses the static `type` property if defined, otherwise converts
   * the class name to snake_case.
   *
   * @returns Type identifier string
   * @protected
   */
  protected getTypeIdentifier(): string {
    const constructor = this.constructor as typeof SerializationMixin;
    return constructor.getTypeIdentifier();
  }

  /**
   * Get the type identifier for this class.
   *
   * Uses the static `type` property if defined, otherwise converts
   * the class name to snake_case.
   *
   * @returns Type identifier string
   * @private
   */
  private static getTypeIdentifier(): string {
    if (this.type) {
      return this.type;
    }

    // Convert class name to snake_case
    return this.name.split(CAMEL_TO_SNAKE_PATTERN).join('_').toLowerCase();
  }
}
